src/expr/src/typecheck.rs

```diff
@@ -0,0 +1,777 @@
+            }
+            MirScalarExpr::CallBinary { expr1, expr2, func } => Ok(func.output_type(
+                expr1.typecheck(source, column_types)?,
+                expr2.typecheck(source, column_types)?,
+            )),
+            MirScalarExpr::CallVariadic { exprs, func } => Ok(func.output_type(
+                exprs
+                    .iter()
+                    .map(|e| e.typecheck(source, column_types))
+                    .collect::<Result<Vec<_>, TypeError>>()?,
+            )),
+            MirScalarExpr::If { cond, then, els } => {
+                let cond_type = cond.typecheck(source, column_types)?;
+
+                // condition must be boolean
+                // ??? does nullability matter? ignoring it here (on purpose)
+                if cond_type.scalar_type != ScalarType::Bool {
+                    return Err(TypeError::MismatchColumn {
+                        source,
+                        got: cond_type,
+                        expected: ColumnType {
+                            scalar_type: ScalarType::Bool,
+                            nullable: true,
+                        },
+                        message: "expected boolean condition".into(),
+                    });
+                }
+
+                let then_type = then.typecheck(source, column_types)?;
+                let else_type = els.typecheck(source, column_types)?;
+                then_type
+                    .union(&else_type)
+                    .map_err(|e| TypeError::MismatchColumn {
+                        source,
+                        got: then_type,
+                        expected: else_type,
+                        message: format!("couldn't compute union of column types for if: {:?}", e),
+                    })
+            }
+        }
+    }
+}
+
+impl AggregateExpr {
+    pub fn typecheck<'a>(
+        &'a self,
+        source: &'a MirRelationExpr,
+        column_types: &[ColumnType],
+    ) -> Result<ColumnType, TypeError<'a>> {
+        let t_in = self.expr.typecheck(source, column_types)?;
+
+        // TODO check that t_in is actually acceptable for `func`
+
+        Ok(self.func.output_type(t_in))
+    }
+}
+
+impl<'a> TypeError<'a> {
+    pub fn source(&self) -> &'a MirRelationExpr {
+        match self {
+            TypeError::Unbound { source, .. }
+            | TypeError::NoSuchColumn { source, .. }
+            | TypeError::MismatchColumn { source, .. }
+            | TypeError::MismatchColumns { source, .. }
+            | TypeError::BadConstantRow { source, .. }
+            | TypeError::BadProject { source, .. }
+            | TypeError::BadTopKGroupKey { source, .. }
+            | TypeError::BadTopKOrdering { source, .. }
+            | TypeError::BadLetRecBindings { source } => source,
+        }
+    }
+}
+
+impl<'a> std::fmt::Display for TypeError<'a> {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
```

Comment:
I _think_ most of our error formatters only use `write!` to not span multiple output lines, maybe we should keep the pattern?

Reply:
It's possible to show output types on one line (though something better than the `Debug` output will probably improve things); it's not really possible to show the source queries on one line, though.

Reply:
The `error!` output here will always be a single line, but the `warn!` output is a more nicely formatted type error message. Happy to change this if long warning lines are too much.